Observer/ConfigChange.php

```diff
@@ -31,39 +34,56 @@ class ConfigChange implements ObserverInterface
      */
     private $typeSenseCollecitons;
 
+    /**
+     * $var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var ConfigService
+     */
+    private ConfigService $configService;
+
+
     /**
      * ConfigChange constructor.
      * @param RequestInterface $request
      */
     public function __construct(
         RequestInterface $request,
         Client $client,
-        AlgoliaHelper $algoliaHelper
+        AlgoliaHelper $algoliaHelper,
+        StoreManagerInterface $storeManager,
+        ConfigService $configService
     ) {
         $this->request = $request;
         $this->typesenseClient = $client->getTypesenseClient();
         $this->algoliaHelper = $algoliaHelper;
         $this->typeSenseCollecitons = $this->typesenseClient->collections;
+        $this->storeManager = $storeManager;
+        $this->configService = $configService;
     }
 
     /**
      * Creates Indexes in Typesense after credentials have been updated
      */
     public function execute(EventObserver $observer)
     {
-        $indexes = $this->algoliaHelper->getIndexDataByStoreIds();
-        unset($indexes[0]); //skip admin store
-
+        $indexes = $this->getMagentoIndexes();
         $existingCollections = $this->getExistingCollections();
-
+        
         foreach($indexes as $index){
-            if(!isset($existingCollections[$index["indexName"]."_products"])){
-                $this->typeSenseCollecitons->create(
-                    [
-                        'name' => $index["indexName"]."_products",
-                        'fields' => [['name' => 'name','type' => 'string']]
-                    ]
-                );
+            $requiredIndexes = ['products','categories', 'pages'];
+            foreach($requiredIndexes as $indexToCreate){
+                if(!isset($existingCollections[$index["indexName"]."_{$indexToCreate}"])){
+                    $this->typeSenseCollecitons->create(
+                        [
+                            'name' => $index["indexName"]."_{$indexToCreate}",
+                            'fields' => [['name' => 'name','type' => 'string'],
+                            ['name' => 'objectID','type' => 'string'],['name' => 'id','type' => 'string']]
```

Comment:
Hi @collymore - 

I think this needs to come from the Algolia Magento config. IE "categories" is required for the product index, as it's used for filtering etc. (see `algoliasearch_products/products/product_additional_attributes` for example) 

I also think we should probably kick this off before indexing starts too, as technically somebody could alter the settings in Typesense or they could change the Algolia config which doesn't have an event listened too. Furthermore, I had to save the correct credentials twice, as the first time the observer seemed to use the old credentials. 

I think "categories" also needs to be a nested field for child/grand child categories. Or we would need to flatten the keys, whichever is easier. 

https://typesense.org/docs/guide/tips-for-searching-common-types-of-data.html#nested-objects

Reply:
// @todo - we need to check if a collection already exists here and if it does use
// @todo $this->typesenseClient->collections[$index["indexName"] . "_{$indexToCreate}"]->update([])
// @todo to just update the schema.
// @see https://typesense.org/docs/0.23.0/api/collections.html#update-or-alter-a-collection